File: exercises/floor02/ex02_branching.py
"""
FLOOR 2 · EXERCISE 2-2 — if / elif / else
Lesson: 2.2

Branching turns booleans into behavior. Order matters: Python takes the
FIRST branch that's true and skips the rest.

Run:  python3 dojo.py check 2-2
Save: python3 dojo.py done 2-2
"""


def classify_roll(roll):
    """A d20 attack roll. Return one of:
        "crit"   for a natural 20
        "miss"   for a natural 1
        "hit"    for anything 11 through 19
        "graze"  for anything 2 through 10
    Use if / elif / else — and think about the ORDER of your checks.
    """
    # TODO
    if roll == 1:
        return "miss"
    elif roll == 20:
        return "crit"
    elif roll in range(2,11):
        return "graze"
    elif roll in range(11,20):
        return "hit"



def door_message(has_key, door_locked):
    """Return the right message:
        locked + no key  -> "The door is locked."
        locked + key     -> "The key turns. The door opens."
        unlocked         -> "The door creaks open."
    """
    # TODO
    if door_locked and not has_key:
        return "The door is locked."
    elif door_locked and has_key:
        return "The key turns. The door opens."
    else:
        return "The door creaks open."

def potion_effect(hp, max_hp):
    """Drinking a potion heals 10, but hp can NEVER exceed max_hp.
    Return the new hp. (This clamp shows up in every game ever made —
    forgetting it is how you get 38/30 HP bugs.)
    """
    # TODO
    hp = hp + 10
    if hp > max_hp:
        hp = max_hp
    return hp

File: exercises/cpp/floor02/ex02_branching.cpp
// FLOOR C2 · EXERCISE c2-2 — if / else if / else
// Lesson: c2.2
//
// Run:  python3 dojo.py check c2-2
// Save: python3 dojo.py done c2-2

#include <string>

// d20 attack roll:
//   20 -> "crit"    1 -> "miss"    11..19 -> "hit"    2..10 -> "graze"
// Order matters — most specific first.
std::string classify_roll(int roll) {
    std::string roll_type = "";
    if(roll == 20){ roll_type = "crit";}
    else if(roll == 1){ roll_type = "miss"; }
    else if(roll >= 11 && roll <= 19){ roll_type = "hit" ;}
    else if(roll >= 2 && roll <= 10){ roll_type = "graze" ;}
    return roll_type;
}

// locked + no key -> "The door is locked."
// locked + key    -> "The key turns. The door opens."
// unlocked        -> "The door creaks open."
std::string door_message(bool has_key, bool door_locked) {
    std::string door_status = "";
    if(!has_key && door_locked) { door_status = "The door is locked.";}
    else if(has_key && door_locked) { door_status = "The key turns. The door opens.";}
    else if(!door_locked) { door_status = "The door creaks open.";}
    return door_status;
}

// Heal 10, but NEVER exceed max_hp — the clamp. Return new hp.
int potion_effect(int hp, int max_hp) {
    
    hp += 10;
    if(hp > max_hp){hp = max_hp;}
    return hp;
}
